sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp

```diff
@@ -37,89 +37,150 @@ RingOutlierFilterComponent::RingOutlierFilterComponent(const rclcpp::NodeOptions
-          std::min(current_pt_distance, next_pt_distance) * distance_ratio_ &&
+        std::max(current_distance, next_distance) <
+          std::min(current_distance, next_distance) * distance_ratio_ &&
         azimuth_diff < 100.f) {
-        continue;
+        continue;  // Determined to be included in the same walk
       }
-      if (isCluster(input_ptr, tmp_indices)) {
-        for (const auto & tmp_idx : tmp_indices) {
-          output_modifier.push_back(
-            std::move(*reinterpret_cast<PointXYZI *>(&input_ptr->data[tmp_idx])));
+
+      if (isCluster(
+            input, std::make_pair(indices[walk_first_idx], indices[walk_last_idx]),
+            walk_last_idx - walk_first_idx + 1)) {
+        for (int i = walk_first_idx; i <= walk_last_idx; i++) {
+          auto output_ptr = reinterpret_cast<PointXYZI *>(&output.data[output_size]);
+          auto input_ptr = reinterpret_cast<const PointXYZI *>(&input->data[indices[i]]);
+
+          if (transform_info.need_transform) {
+            Eigen::Vector4f p(input_ptr->x, input_ptr->y, input_ptr->z, 1);
+            p = transform_info.eigen_transform * p;
+            output_ptr->x = p[0];
+            output_ptr->y = p[1];
+            output_ptr->z = p[2];
+            output_ptr->intensity = input_ptr->intensity;
+          } else {
+            *output_ptr = *input_ptr;
+          }
+
+          output_size += output.point_step;
         }
       }
-      tmp_indices.clear();
-    }
-    if (tmp_indices.empty()) {
-      continue;
+
+      walk_first_idx = idx + 1;
     }
-    if (isCluster(input_ptr, tmp_indices)) {
-      for (const auto & tmp_idx : tmp_indices) {
-        output_modifier.push_back(
-          std::move(*reinterpret_cast<PointXYZI *>(&input_ptr->data[tmp_idx])));
+
+    if (walk_first_idx > walk_last_idx) continue;
+
+    if (isCluster(
+          input, std::make_pair(indices[walk_first_idx], indices[walk_last_idx]),
+          walk_last_idx - walk_first_idx + 1)) {
+      for (int i = walk_first_idx; i <= walk_last_idx; i++) {
+        auto output_ptr = reinterpret_cast<PointXYZI *>(&output.data[output_size]);
+        auto input_ptr = reinterpret_cast<const PointXYZI *>(&input->data[indices[i]]);
+
+        if (transform_info.need_transform) {
+          Eigen::Vector4f p(input_ptr->x, input_ptr->y, input_ptr->z, 1);
+          p = transform_info.eigen_transform * p;
+          output_ptr->x = p[0];
+          output_ptr->y = p[1];
+          output_ptr->z = p[2];
+          output_ptr->intensity = input_ptr->intensity;
+        } else {
+          *output_ptr = *input_ptr;
+        }
+
+        output_size += output.point_step;
       }
     }
-    tmp_indices.clear();
   }
+
+  output.data.resize(output_size);
+
+  // Note that `input->header.frame_id` is data before converted when `transform_info.need_transform
+  // == true`
+  output.header.frame_id = !tf_input_frame_.empty() ? tf_input_frame_ : tf_input_orig_frame_;
```

Comment:
Since this outlier filter algorithm works in "tf_input_frame_" coordinates, the output pointcloud at this point is in the same  "tf_input_frame_" coordinates.
```suggestion
  output.header.frame_id = tf_input_frame_;
```